internal/impl/opensearch/output.go

```diff
@@ -0,0 +1,454 @@
+	}
+
+	if conf.TLS.Enabled {
+		var err error
+		if e.tlsConf, err = conf.TLS.Get(); err != nil {
+			return nil, err
+		}
+	}
+	return &e, nil
+}
+
+//------------------------------------------------------------------------------
+
+// Connect attempts to establish a connection to a OpenSearch broker.
+func (e *OpenSearch) Connect(ctx context.Context) error {
+	if e.client != nil {
+		return nil
+	}
+	boff := e.backoffCtor()
+	opts := os.Config{
+		Addresses:     e.urls,
+		RetryOnStatus: []int{502, 503, 504, 429},
+		RetryBackoff: func(attempt int) time.Duration {
+			if attempt == 1 {
+				boff.Reset()
+			}
+			return boff.NextBackOff()
+		},
+		MaxRetries: int(e.conf.MaxRetries),
+	}
+	if e.conf.Auth.Enabled {
+		opts.Username = e.conf.Auth.Username
+		opts.Password = e.conf.Auth.Password
+	}
+
+	opts.Transport = &http.Transport{
+		ResponseHeaderTimeout: e.timeout,
+	}
+	if e.conf.TLS.Enabled {
+		opts.Transport = &http.Transport{
+			TLSClientConfig:       e.tlsConf,
+			ResponseHeaderTimeout: e.timeout,
+		}
+
+	}
+
+	awsOpts, err := AWSOptFn(opts.Transport, e.conf)
+	if err != nil {
+		return err
+	}
+	if awsOpts != nil {
+		opts.Transport = awsOpts
+
+	}
+
+	if e.conf.GzipCompression {
+		opts.CompressRequestBody = true
+	}
+
+	client, err := os.NewClient(opts)
+	if err != nil {
+		return err
+	}
+
+	e.client = client
+	e.log.Infof("Sending messages to OpenSearch index at urls: %s\n", e.urls)
+	return nil
+}
+
+type pendingBulkIndex struct {
+	Action   string
+	Index    string
+	Pipeline string
+	Routing  string
+	Type     string
+	Doc      interface{}
+	ID       string
+}
+
+// WriteBatch will attempt to write a message to OpenSearch, wait for
+// acknowledgement, and returns an error if applicable.
+func (e *OpenSearch) WriteBatch(ctx context.Context, msg message.Batch) error {
+	return e.Write(msg)
+}
+
+// Write will attempt to write a message to OpenSearch, wait for
+// acknowledgement, and returns an error if applicable.
+func (e *OpenSearch) Write(msg message.Batch) error {
+	if e.client == nil {
+		return component.ErrNotConnected
+	}
+
+	requests := make([]*pendingBulkIndex, msg.Len())
+	if err := msg.Iter(func(i int, part *message.Part) error {
+		jObj, ierr := part.AsStructured()
+		if ierr != nil {
+			e.log.Errorf("Failed to marshal message into JSON document: %v\n", ierr)
+			return fmt.Errorf("failed to marshal message into JSON document: %w", ierr)
+		}
+		requests[i] = &pendingBulkIndex{
+			Action:   e.actionStr.String(i, msg),
+			Index:    e.indexStr.String(i, msg),
+			Pipeline: e.pipelineStr.String(i, msg),
+			Routing:  e.routingStr.String(i, msg),
+			Type:     e.typeStr.String(i, msg),
+			Doc:      jObj,
+			ID:       e.idStr.String(i, msg),
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+	start := time.Now()
+	b, _ := opensearchutil.NewBulkIndexer(opensearchutil.BulkIndexerConfig{
+		FlushInterval: time.Second * 1,
+		FlushBytes:    0,
+		Client:        e.client,
+	})
+	ctx := context.Background()
```

Comment:
I think we'll want to propagate the `ctx` parameter from `WriteBatch` here. Otherwise, `b.Add` and `b.Close` could block forever...

Reply:
Totally Agreed!